libraries/core_libs/consensus/src/rewards/rewards_stats.cpp

```diff
@@ -58,26 +58,47 @@ BlockStats Stats::getBlockStats(const PeriodData& blk, const std::vector<gas_t>&
 
 std::vector<BlockStats> Stats::processStats(const PeriodData& current_blk, const std::vector<gas_t>& trxs_gas_used,
                                             DbStorage::Batch& write_batch) {
-  std::vector<BlockStats> res;
   const auto current_period = current_blk.pbft_blk->getPeriod();
   const auto frequency = getCurrentDistributionFrequency(current_period);
   auto block_stats = getBlockStats(current_blk, trxs_gas_used);
+
   // Distribute rewards every block
   if (frequency == 1) {
     return {block_stats};
   }
 
-  blocks_stats_.emplace(current_period, std::move(block_stats));
+  blocks_stats_.emplace(current_period, block_stats);
   // Blocks between distribution. Process and save for future processing
   if (current_period % frequency != 0) {
     // Save to db, so in case of restart data could be just loaded for the period
     saveBlockStats(current_period, blocks_stats_[current_period], write_batch);
     return {};
   }
 
-  res.reserve(blocks_stats_.size());
-  std::transform(blocks_stats_.begin(), blocks_stats_.end(), std::back_inserter(res),
-                 [](auto& t) { return std::move(t.second); });
+  // Transform ordered (or unordered) blocks stats into vector
+  auto transformStatsToVector = [](const auto& blocks_stats) {
+    std::vector<BlockStats> stats_vec;
+    stats_vec.reserve(blocks_stats.size());
+
+    std::transform(blocks_stats.begin(), blocks_stats.end(), std::back_inserter(stats_vec),
+                   [](const auto& t) { return t.second; });
+    return stats_vec;
+  };
+
+  std::vector<BlockStats> res;
+
+  // Blocks stats were not sorted by period before aspen hardfork part one
+  if (current_period < kHardforksConfig.aspen_hf.block_num_part_one) {
+    res = transformStatsToVector(blocks_stats_);
+  } else {
+    // Blocks stats are sorted by period after aspen hardfork part one
+    std::map<PbftPeriod, BlockStats> ordered_blocks_stats;
```

Comment:
if you do a double insertion the number of elements in the map will not match "rewards_distribution_frequency". I am not that strong about using a vector but it would be more efficient